### PR TITLE
Fix #99: phase timing and operation-count instrumentation

### LINKED ISSUE #99: [enh] Phase timing and operation-count instrumentation ("how long did Scalpel take and where did it go?")
## Problem

No timing or counting instrumentation exists anywhere in the analysis path (zero uses of nanoTime/currentTimeMillis). For a tool whose entire value proposition is time saved, a user cannot answer "how long did Scalpel take and where did it go?", and no performance claim can be confirmed or refuted in the field.

## Proposal

A small `Timings` helper (plain `System.nanoTime()` deltas, no new dependency) covering repoOpen, fetch, mergeBase, diff, status, readOldPoms, moduleMapping, pomAnalysis, transitiveResolve, trim and applySkipTests, with one INFO summary line at the end, plus `timings` and `operations` objects in the JSON report.

The operation counts matter more than the milliseconds: they are machine-independent, so they can be asserted in CI without flaking. E.g. a count-based gate like "a single changed POM in a 500-module reactor performs at most N git blob reads" catches redundant-read regressions the day they are written. Useful counts: git blobs read, POM files parsed, dependency resolutions performed, resolve cache hits/misses, resource files opened, bytes scanned.

This would also be the foundation for a shadow/verification mode measuring real savings per build. Happy to PR the `Timings` helper first if welcome.

### PR DESCRIPTION
**Pull request by paoloantinori**
## Problem

As filed in #99: zero uses of nanoTime/currentTimeMillis existed anywhere in the analysis path. For a tool whose value proposition is time saved, a user could not answer "how long did Scalpel take and where did it go?", and no performance claim could be confirmed or refuted in the field.

## Change

- A small `Timings` collector in core (plain nanoTime deltas, no new dependencies): named phases (`start/stop`) accumulating millis, named operation counters, formatting helpers.
- **Phases**, named after what the current code does: `repoOpen, fetch, mergeBase, diff, status, readOldPoms` (in `ScalpelCore`) and `moduleMapping, pomAnalysis, transitiveResolve, trim, applySkipTests` (in the participant). Every start/stop pair is lexically adjacent with the stop in a finally.
- **Operation counts**: git blobs read (equals the old-POM map size: one read per entry), effective models built, dependency resolutions performed + cache hits (hits and misses mutually consistent on both caches), resource entries visited by the walk (accumulated on every exit path including budget/depth termination).
- **One INFO summary line** emitted in a finally so it also appears on bail-outs: `Scalpel: analysis took 113ms (repoOpen=89ms, mergeBase=11ms, ...) ops: blobs=1, models=6, resolves=4, resolveCacheHits=2, resources=2`.
- **Report**: optional `timings` ({totalMillis, phases}) and `operations` objects, emitted independently, omitted entirely when empty; status-only documents never carry them. Schema v2 declares both (no version bump, per policy); README example and version history updated.
- `detectChanges` gains a 4-arg overload taking the collector; the published 3-arg signature delegates unchanged (null collector fails fast with a clear message).

## Verification

TDD: RED first (4 targeted failures: no summary line, no report fields, schema lacking the properties), then GREEN. Suites green (core 184, extension3 177), plus real Maven IT runs showing the summary line with live numbers and the report carrying/omitting the objects correctly. Known coordination note: PR #157's schema equality guard (still open) will need `timings`/`operations` added to its emittable set when both meet in main.

## Summary by CodeRabbit

* **New Features**
  * Analysis reports now optionally include phase timings and operation counts.
  * Added visibility into analysis duration, processing phases, dependency resolutions, cache hits, Git reads, and scanned resources.
  * Added a concise analysis timing summary to logs.
  * Updated report documentation and schema to describe the new fields.

* **Bug Fixes**
  * Reports omit instrumentation fields when no timing or operation data was recorded.

### CODE REVIEW

File: core/src/main/java/eu/maveniverse/maven/scalpel/core/ScalpelReport.java [hunk 1]
```diff
@@ -328,6 +334,38 @@ public String toJson() {
...
+            sb.append("  \"operations\": {");
+            boolean firstOperation = true;
+            for (String operation : timings.getOperationNames()) {
+                if (!firstOperation) {
+                    sb.append(", ");
+                }
+                firstOperation = false;
+                sb.append(jsonString(operation)).append(": ").append(timings.getOperationCount(operation));
+            }
+            sb.append("}\n");
```
Comment (gnodet) [inline core/src/main/java/eu/maveniverse/maven/scalpel/core/ScalpelReport.java]: Minor formatting inconsistency: the operations block closes with `sb.append("}\n")` (trailing newline), but the timings block closes with `sb.append("  }")` (no trailing newline). Since the root close is `sb.append("\n}\n")`, the operations path produces a double-newline (`}\n\n}\n` — a blank line in the output) when operations is the last emitted field.

Dropping the `\n` here would make it consistent with the timings pattern:

```suggestion
            sb.append("}");
```

**Review by gnodet — APPROVED** [review]
Thorough, well-tested instrumentation that adds phase timing and operation counting with zero new dependencies. Clean API extension preserving backward compatibility, disciplined finally-block placement, and TDD approach. Two minor confirmed observations noted inline.

**Strengths worth highlighting:**
- The `Timings` class is elegantly designed — zero new dependencies, `nanoTime` deltas accumulated via `LinkedHashMap` for first-seen ordering, a running-phase-aware `getPhaseMillis` that reports elapsed-so-far even before `stop`, and a safe-from-finally `stop` that is a no-op when the phase was never started.
- The TDD commit sequence (RED → GREEN → review fix) is clearly visible and makes the design intent very readable.
- Every `start`/`stop` pair is lexically adjacent with `stop` in a `finally`, which is the right pattern for instrumentation that should never leak or corrupt on exception paths.
- The `timings` and `operations` JSON objects are independently conditional (each emitted only when non-empty), which keeps status-only reports and uninstrumented paths clean.
- Comprehensive test coverage: unit tests for `Timings`, serialization tests for the report, schema validation tests, and full lifecycle-participant integration tests including the console INFO summary line.

**Coordination note:** PR #157 (schema drift guard) will need `timings`/`operations` added to its emittable set when both meet in `main` — well-documented in the PR body.

### 📋 PR Metadata

| Aspect | Current | Suggested |
|--------|---------|----------|
| Category | `bug` | `enhancement` (feature) |
| Labels | `bug`, `documentation`, `java` | + `enhancement`, − `bug` |
| Milestone | _(none)_ | `0.3.11` |

> Issue #99 is titled "[enh] Phase timing and operation-count instrumentation" — this is a new observability feature, not a bug fix. The `bug` label appears to be a misclassification from the autolabeler matching "fix" in the branch name.

_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of Guillaume Nodet_

**Review by gnodet — APPROVED** [review]
Well-designed phase timing and operation-count instrumentation that adds genuine observability value. Code is clean, tests are thorough, and CI is green on all matrix combinations. Two low-severity cosmetic observations:

1. **JSON formatting inconsistency:** The `operations` object closing uses `sb.append("}\n")` (with trailing newline) while the `timings` object closing uses `sb.append("  }")` (without). When `operations` is the last field before the root `sb.append("\n}\n")`, this produces a double-newline (blank line) before the closing brace. Changing to `sb.append("}")` would make it consistent.

2. **Schema tightening:** The `phases` property in the schema is declared as `"type": "object"` without `"additionalProperties": {"type": "integer", "minimum": 0}`. Since phase names are dynamic ("may grow additively"), `additionalProperties` is the right JSON Schema mechanism to constrain value types. The sibling `totalMillis` already uses `{"type": "integer", "minimum": 0}`.

### 📋 PR Metadata

| Aspect | Current | Suggested |
|--------|---------|-----------|
| Category | `bug` | `enhancement` (issue #99 is titled "[enh]") |
| Labels | `bug`, `documentation`, `java` | replace `bug` with `enhancement` |
| Milestone | _(none)_ | `0.4.0` |

_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of Guillaume Nodet_

**Review by gnodet — APPROVED** [review]
Re-reviewed after fix commit 4aed60b. All prior findings addressed:

- **Negative `totalMillis` validation:** `Builder.timings()` now rejects `totalMillis < 0` with `IllegalArgumentException`, consistent with the schema's `minimum: 0`.
- **Phase ordering fix:** A dedicated `phaseOrder` `LinkedHashSet` records start order, so overlapping phases (start outer → start inner → stop inner) correctly report `[outer, inner]` instead of stopped-first order. Regression test covers this.
- **Schema `additionalProperties`:** The `phases` object now declares `additionalProperties: {type: integer, minimum: 0}`, rejecting invalid values like `{"diff": -1}` or `{"diff": "invalid"}`. The in-test schema validator was extended to support `additionalProperties` validation.
- **Operations JSON formatting:** The trailing `\n` on the operations closing brace was removed, eliminating the double-newline before the root closing brace.

The `Timings` class is well-designed — zero dependencies, accumulating nanoTime deltas with LinkedHashMap ordering, running-phase-aware reporting, and safe no-op `stop()` for finally blocks. Test coverage is thorough across all edge cases.

_This review was generated by an AI agent, Hermès on behalf of @gnodet._
